bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/handler/ADBridgeHandler.java

```diff
@@ -0,0 +1,428 @@
+        logger.debug("Sending AD command: {}", command);
+        try {
+            if (writer != null) {
+                writer.write(command.toString());
+                writer.flush();
+            }
+        } catch (IOException e) {
+            logger.info("Exception while sending command: {}", e.getMessage());
+            writeException = true;
+        }
+    }
+
+    protected abstract void connect();
+
+    protected abstract void disconnect();
+
+    protected void scheduleConnectRetry(long waitMinutes) {
+        logger.debug("Scheduling connection retry in {} minutes", waitMinutes);
+        connectRetryJob = scheduler.schedule(this::connect, waitMinutes, TimeUnit.MINUTES);
+    }
+
+    protected void startMsgReader() {
+        synchronized (msgReaderThreadLock) {
+            msgReaderThread = new Thread(this::readerThread, "AD Reader");
+            msgReaderThread.setDaemon(true);
+            msgReaderThread.start();
+        }
+    }
+
+    protected void stopMsgReader() {
+        synchronized (msgReaderThreadLock) {
+            if (msgReaderThread != null) {
+                logger.trace("Stopping reader thread.");
+                msgReaderThread.interrupt();
+                msgReaderThread = null;
+            }
+        }
+    }
+
+    /**
+     * Method executed by message reader thread
+     */
+    private void readerThread() {
+        logger.debug("Message reader thread started");
+        String msg = null;
+        try {
+            // Send version command to get device to respond with VER message.
+            sendADCommand(ADCommand.getVersion());
+            while (!Thread.interrupted() && reader != null && (msg = reader.readLine()) != null) {
+                logger.trace("Received msg: {}", msg);
+                ADMsgType mt = ADMsgType.getMsgType(msg);
+                if (mt != ADMsgType.INVALID) {
+                    lastReceivedTime = new Date();
+                }
+                try {
+                    switch (mt) {
+                        case KPM:
+                            parseKeypadMessage(msg);
+                            break;
+                        case REL:
+                        case EXP:
+                            parseRelayOrExpanderMessage(mt, msg);
+                            break;
+                        case RFX:
+                            parseRFMessage(msg);
+                            break;
+                        case LRR:
+                            parseLRRMessage(msg);
+                            break;
+                        case VER:
+                            parseVersionMessage(msg);
+                            break;
+                        case INVALID:
+                        default:
+                            break;
+                    }
+                } catch (MessageParseException e) {
+                    logger.info("Error {} while parsing message {}", e.getMessage(), msg);
```

Comment:
Well, at least for now, I would really like users to be warned of any parse errors so that they can report the problem. Otherwise there may be silent failures in some cases which we won't know about. Do you think it would make sense to leave it at INFO for now and bump it down to DEBUG later? Or I can bump it up to WARN for now if you would rather reserve INFO level.